### ISSUE TITLE
Boost locale is built without ICU backend

### ISSUE THREAD

**Issue by dimztimz** [Homebrew/homebrew-core#23288]
**Please note we will close your issue without comment if you delete, do not read or do not fill out the issue checklist below and provide ALL the requested information. If you repeatedly fail to use the issue template, we will block you from ever submitting issues to Homebrew again.**

# Please always follow these steps:
- [x] Confirmed this is a problem with `brew install`ing one, specific Homebrew/homebrew-core formula (not cask or tap) and not every time you run `brew`? If it's a general `brew` problem please file this issue at https://github.com/Homebrew/brew/issues/new. If it's a `brew cask` problem please file this issue at https://github.com/caskroom/homebrew-cask/issues/new. If it's a tap (e.g. Homebrew/homebrew-php) problem please file this issue at the tap.
- [x] Ran `brew update` and retried your prior step?
- [x] Ran `brew doctor`, fixed all issues and retried your prior step?
- [ ] Ran `brew gist-logs <formula>` (where `<formula>` is the name of the formula that failed) and included the output link?
- [ ] If `brew gist-logs` didn't work: ran `brew config` and `brew doctor` and included their output with your issue?

The package "boost" is a collection of C++ libraries. Among them one is "boost locale" which provides facilities for internationalization with a modern C++ interface. The library optionally depends on "icu4c". When icu4c is not present boost locale will compile, but will provide limited functionality. Nobody really uses boost locale without icu4c.

To make it work with icu, icu4c has to be installed and the boost build system will detect the library and nothing specially has to be done. Boost currently CAN be installed with icu4c like this:

    brew install boost --with-icu4c

But takes about half an hour to compile. So my vote is to make "icu4c" mandatory and provide built binaries with icu4c support. Or maybe split boost locale from the basic package and make the boost-locale depend on icu4c.

Linux distributions ([Ubuntu](https://packages.ubuntu.com/artful/libboost-locale1.62.0)) provide boost locale with icu out of the box.

**Comment by fxcoudert** [comment]
Analytics says right now `icu4c` is used in 0.46% of boost installs, 0.73% of user-requested installs.

**Comment by dimztimz** [comment]
Not sure what that means.

Not everyone uses boost-locale from boost, but when "boost-locale" is used, icu4c is a necessity.

**Comment by ilovezfs** [comment]
@dimztimz Unfortunately, given how infrequently this option is used, we'll need to pass on both ideas for now. If it becomes more popular, we can reconsider making it mandatory or having the separate formula.

**Comment by dimztimz** [comment]
Ok you are basing the conclusion on false statistics.

Boost has like 50 libraries. ICU is used only in 2, boost-locale and boost-regex. So, people using the other 48 libraries don't need ICU, but people that use the 2, most likely need ICU.

I mentioned that Ubuntu and Debian build WITH ICU. I'd mention Arch Linux which has somewhat less granular packages, and ships boost in 2 parts, package [boost-libs](https://www.archlinux.org/packages/extra/x86_64/boost-libs/) that ships precompiled shared libraries and pacakge [boost](https://www.archlinux.org/packages/extra/x86_64/boost/) that ships headers (and header-only libraries). ICU is a dependency on boost-libs. [FreeBsd](https://www.freebsd.org/cgi/ports.cgi?query=boost-libs&stype=all&sektion=all) also builds boost with ICU.

Those 2 libraries in a large amount of use cases are unusable without ICU.

So basically, all major Linux distributions, and even BSD OS provide Boost with ICU. Only homebrew does not.

**Comment by fxcoudert** [comment]
We have 110 formulas in Homebrew that depend on boost, and apparently none requires ICU. And we do provide ICU, optionally, meaning it forces a build from source (but it does work).

The choice we have is between:
1. leaving as is: people who need boost with ICU can get it, but it will build from source
2. making ICU mandatory: this means that our binaries become non-relocatable, so all users who installed Homebrew anywhere other than `/usr/local` will have boost built from source

We think that option 1 is the one that minimises the annoyance of our user base.

**Comment by dimztimz** [comment]
It is annoyance to offer a disfunctional library too. I will say again. Boost locale without ICU is almost useless.

I am willing to make a pull request in the future that splits boost locale from the main package and has mandadory icu dependency. Just like to get assurance that it will get accepted.

**Comment by ilovezfs** [comment]
@dimztimz as I already said above, we've decided against the idea of a separate formula for now.
